src/references/referenceService.ts

```diff
@@ -0,0 +1,230 @@
+import { IFormattedText, IReferenceInstance, IReference } from "./reference";
+
+export class referenceService {
+  constructor(
+    // Collections of references
+    private referenceLibrary = new Map(),
+    private referenceInstanceLibrary = new Map(),
+    private arrayOfRefsForFormatting: number[] = [],
+    // Trackers
+    private refInstanceCount: number = 0,
+    private lastRefToBeFormatted: number = 0,
+    // Output strings
+    private finalText: string = "",
+    private inputText: string = "",
+    // Regex
+    private REFERENCE_IDENTIFICATION_REGEX: RegExp = /\[(.+?)\]/g
+  ) {}
+
+  public post(inputBody: string): IFormattedText {
+    this.inputText = inputBody;
+    const output = this.refFormatter(this.inputText);
+    return {
+      formattedText: output.text,
+      references: output.referenceList,
+    };
+  }
+
+  refFormatter(inputText: string) {
+    let remainingText: string = "";
+    let instanceCounter = 1;
+
+    // Track where each reference occurs
+    const refInstance = this.identifyReferences(inputText);
+    for (const instance of refInstance) {
+      const newRefInstance: IReferenceInstance = {
+        nameYear: instance[0],
+        index: instance.index!,
+      };
+      this.referenceInstanceLibrary.set(instanceCounter, newRefInstance);
+      instanceCounter++;
+    }
+
+    // Build a collection of reference authors, indexes of their references, and length of each reference
+    // Checks to see if there's a reference directly after the one currently being focused on
+    const rawRefs = this.identifyReferences(inputText);
+    for (const rawRef of rawRefs) {
+      if (this.recordExists(rawRef)) {
+        const existingRef = this.referenceLibrary.get(rawRef[0]);
+        const existingLocations = existingRef.locations;
+        const newLocation = rawRef.index;
+        existingLocations.push(newLocation);
+        this.arrayOfRefsForFormatting.push(existingRef.referenceId);
+      } else if (!this.recordExists(rawRef)) {
+        const newRef: IReference = {
+          referenceId: this.referenceLibrary.size + 1,
+          locations: [rawRef.index!],
+          referenceLength: rawRef[0].length,
+        };
+        this.referenceLibrary.set(rawRef[0], newRef);
+        this.arrayOfRefsForFormatting.push(newRef.referenceId);
+      }
+      remainingText = inputText.slice(rawRef.index! + rawRef[0].length + 1);
+
+      // Check text immediately after the current reference. If nothing adjacent, format existing collection of references.
+      this.identifyAdjacentReferences(remainingText);
+      this.refInstanceCount++;
+    }
+
+    const referencesLeftToFormat =
+      this.arrayOfRefsForFormatting.length > 0 ? true : false;
+
+    if (referencesLeftToFormat) {
+      this.concatReferencesToFinalOutput();
+    }
+    let referenceList = this.createReferenceList(this.referenceLibrary);
+    return { text: this.finalText, referenceList: referenceList };
```

Comment:
```suggestion
    return { text: this.finalText, referenceList };
```

https://ui.dev/shorthand-properties

Reply:
Neat